### ISSUE TITLE
BlankLineBeforeDeclarationRule: Consecutive function declarations using expression bodies should not be flagged

### ISSUE THREAD

**Issue by 3flex** [pinterest/ktlint#2109]
## Expected Behavior
Consecutive function declarations using expression bodies should not be flagged by BlankLineBeforeDeclarationRule.

## Observed Behavior
The functions are flagged in this sample:

```kotlin
object TestObject {
    val findings: Map<String, List<String>> = emptyMap()
    val notifications: Collection<String> = emptyList()
    val metrics: Collection<String> = emptyList()

    fun <V> getData(key: Key<V>): V = throw UnsupportedOperationException()
    fun <V> addData(key: Key<V>, value: V): Nothing = throw UnsupportedOperationException()
    fun add(projectMetric: String): Nothing = throw UnsupportedOperationException()
}
```

These seem similar enough to consecutive property declarations that they shouldn't be flagged by the rule.

## Steps to Reproduce
Check code above with BlankLineBeforeDeclarationRule.

## Your Environment
* Version of ktlint used: 0.50.0
* Relevant parts of the `.editorconfig` settings: N/A
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): Using detekt wrapper for the rule
* Version of Gradle used (if applicable): 8.2
* Operating System and version: Windows 11

**Comment by paul-dingemans** [comment]
The purpose of the `blank-line-before-declaration` rule is to insert blank lines before any declaration. The idea is to increase the visibility of the declarations. Exceptions on this rule are:
* If a declaration is the first code element in a block, it is not preceded by a blank line
* Consecutive properties do not require blank lines as it is not uncommon that properties sometime are more strongely related and grouping them together makes the code more readable.

Although the latter argument can also be applied to functions (or classes) it is less common. Also, it is widely accepted practice to have a blank line before a function.

Treating functions in the same way as properties as you suggest is almost identically to disabling the rule entirely.

**Comment by 3flex** [comment]
> Treating functions in the same way as properties as you suggest is almost identically to disabling the rule entirely.

I'm only suggesting that consecutive functions that use an expression body are ignored, not all function declarations. I don't think this is the same as disabling the rule entirely? Functions declared with block bodies would still be flagged.

**Comment by paul-dingemans** [comment]
> I'm only suggesting that consecutive functions that use an expression body are ignored, not all function declarations. I don't think this is the same as disabling the rule entirely? Functions declared with block bodies would still be flagged.

This was not clear to me from your original post. An you're right, that it is not the same as disabling the rule entirely.

But still I don't think it would be wise to do so. [The Kotlin Coding conventions prefer the usage of an expression body over block bodies](https://kotlinlang.org/docs/coding-conventions.html#functions). If the expression is a multiline expression, the blank line will still add clarity to separate the functions. 

Without blank lines before functions:
```
fun foo1() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }
fun bar() = foo1() + foo2()
fun foo2() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }
```

With blank lines before functions:
```
fun foo1() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }

fun bar() = foo1() + foo2()

fun foo2() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }
```

Not forcing a blank line before a single line function with an expression body would result in:
```
fun foo1() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }
fun bar() = foo1() + foo2()

fun foo2() =
    if (foo()) {
       "fooooooooooooooooooooooooooooooo"
    } else {
        "bar"
   }
```
[Of course that can be fixed by changing the `blank-line-before-declaration` into something like `blank-lines-around-declaration`]

**Comment by 3flex** [comment]
> This was not clear to me from your original post.

Sorry about that. Do you have a suggestion for how to make the original post clearer? I only referred to expression bodies in the title, description and example.

> If the expression is a multiline expression, the blank line will still add clarity

I agree - it just seems a little clunky for the functions with expression bodies that all fit on one line to be separated, but then if that's an exception to the rule, it may be come harder to explain what the intent of the rule is, when it applies and when it doesn't. For now though I don't think this is one that we'll enable.
